src/java.base/share/classes/java/lang/Enum.java

```diff
@@ -166,13 +168,29 @@ public final boolean equals(Object other) {
         return this==other;
     }
 
+    /**
+     * The hash code of this enumeration constant.
+     *
+     * @implNote Once initialized, the field value does not change.
+     * Hotspot's identity hash code generation also never returns zero
+     * as the identity hash code. This allows to treat zero as the marker
+     * for the un-initialized value for both {#link @Stable} and the lazy
+     * initialization code.
+     */
+    @Stable
+    private int hash;
+
     /**
      * Returns a hash code for this enum constant.
      *
      * @return a hash code for this enum constant.
      */
     public final int hashCode() {
-        return super.hashCode();
+        int hc = hash;
+        if (hc == 0) {
+            hc = hash = System.identityHashCode(this);
```

Comment:
Why not `hc = hash = super.hashCode()`?

Reply:
Shouldn’t `invokespecial` produce a non‑virtual call as well?

And `this`/`super` can never be `null`.

Reply:
`super.hashCode()` is a virtual call. `System.identityHashCode` is the static call. I don't understand where `invokespecial` enters the picture here.

Reply:
`invokespecial` is used to call instance methods in a non‑virtual manner.

--------------------------------------------------------------------------------

Using `super.hashCode()` in `java.lang.Enum` produces the bytecode:
```
aload_0;
invokespecial java/lang/Object.hashCode:()I;
```

Reply:
I see. It does not matter here, because for performance, we largely only care about the distinction between virtual/static cases in C2 intrinsic for hashcode.

Reply:
Also fyi invokespecial now can be virtual calls (mainly calling private methods in the nest) since introduction of nestmates.